File: TemplateCS/Addin/AddinCommandBars.cs
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Runtime.InteropServices;
using Microsoft.Office.Core;
using stdole;
using System.Windows.Forms;
using System.Drawing.Imaging;

namespace $csprojectname$
{
    /// <summary>
    /// User interface manager for Visio 2007 and below 
    /// (works also with latest versions, but makes no sense there, as there is ribbon)
    /// Creates and controls a custom command bar with buttons
    /// </summary>
    /// 
    public partial class AddinUI
    {
        private string _toolbarName;

        private readonly List<string> _commands =
            new List<string>();

        /// <summary>
        /// Constructs the UI manager
        /// </summary>
        /// <param name="toolbarName">The name of the toolbar to install</param>
        /// <param name="commands">The list of buttons to add (list of button ids)</param>
        public void StartupCommandBars(string toolbarName, IEnumerable<string> commands)
        {
            _toolbarName = toolbarName;
            _commands.AddRange(commands);

            ThisAddIn.Application.VisioIsIdle += ApplicationIdle;
            UpdateCommandBars();
        }

        public void ShutdownCommandBars()
        {
            ThisAddIn.Application.VisioIsIdle -= ApplicationIdle;
        }

        private bool _updateRequest;

        internal void UpdateCommandBars()
        {
            _updateRequest = true;
        }

        void ApplicationIdle(object app)
        {
            if (!_updateRequest)
                return;

            _updateRequest = false;

            UpdateToolbar();
        }

        private readonly Dictionary<string, CommandBarButton> _buttons =
            new Dictionary<string, CommandBarButton>();

        private static CommandBar FindCommandBar(CommandBars cbs, string name)
        {
            return cbs.Cast<CommandBar>().FirstOrDefault(cb => cb.Name == name);
        }

        private void UpdateToolbar()
        {
            var cbs = (CommandBars)ThisAddIn.Application.CommandBars;

            var cb = FindCommandBar(cbs, _toolbarName) ?? cbs.Add(_toolbarName);
            cb.Visible = true;

            foreach (var id in _commands)
                InstallButton(cb, id);
        }

        private void InstallButton(CommandBar cb, string id)
        {
            CommandBarButton thisButton;
            _buttons.TryGetValue(id, out thisButton);

            // Recreate the button, otherwise it's state may be broken in Visio in some cases
            if (thisButton != null)
            {
                thisButton.Click -= CommandBarButtonClicked;
                _buttons.Remove(id);
                Marshal.ReleaseComObject(thisButton);
            }

            var button = (CommandBarButton)cb.FindControl(Tag: id) ??
                (CommandBarButton)cb.Controls.Add(MsoControlType.msoControlButton);

            button.Enabled = ThisAddIn.IsCommandEnabled(id);

            var checkState = ThisAddIn.IsCommandChecked(id);
            button.State = checkState ? MsoButtonState.msoButtonDown : MsoButtonState.msoButtonUp;

            button.Tag = id;
            button.Caption = ThisAddIn.GetCommandLabel(id);
            SetCommandBarButtonImage(button, id);

            button.Click += CommandBarButtonClicked;

            _buttons.Add(id, button);
        }

        private void CommandBarButtonClicked(CommandBarButton ctrl, ref bool cancelDefault)
        {
            ThisAddIn.OnCommand(ctrl.Tag);
        }

        private void SetCommandBarButtonImage(CommandBarButton button, string id)
        {
            var image = ThisAddIn.GetCommandBitmap(id + "_sm");
            if (image == null)
                return;

            Bitmap picture, mask;
            BitmapToPictureAndMask(image, out picture, out mask);

            button.Picture = PictureConvert.ImageToPictureDisp(picture);
            button.Mask = PictureConvert.ImageToPictureDisp(mask);
        }

        static public void BitmapToPictureAndMask(Bitmap bm, out Bitmap picture, out Bitmap mask)
        {
            var w = bm.Width;
            var h = bm.Height;

            var pictureData = new byte[3 * w * h];
            var maskData = new byte[3 * w * h];

            var bmBits = bm.LockBits(new Rectangle(0, 0, w, h), ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);
            var bits = new byte[4 * w * h];
            Marshal.Copy(bmBits.Scan0, bits, 0, 4 * w * h);
            bm.UnlockBits(bmBits);

            for (var y = 0; y < h; ++y)
            {
                for (var x = 0; x < w; ++x)
                {
                    var srcIdx = (x + y * w) * 4;
                    var dstIdx = (x + y * w) * 3;

                    pictureData[dstIdx + 0] = bits[srcIdx + 0];
                    pictureData[dstIdx + 1] = bits[srcIdx + 1];
                    pictureData[dstIdx + 2] = bits[srcIdx + 2];

                    var t = (bits[srcIdx + 3] < 128) ? (byte)255 : (byte)0;

                    maskData[dstIdx + 0] = t;
                    maskData[dstIdx + 1] = t;
                    maskData[dstIdx + 2] = t;
                }
            }

            var rect = new Rectangle(0, 0, w, h);

            picture = new Bitmap(w, h, PixelFormat.Format24bppRgb);
            var pictureBits = picture.LockBits(rect, ImageLockMode.WriteOnly, picture.PixelFormat);
            Marshal.Copy(pictureData, 0, pictureBits.Scan0, w * h * 3);
            picture.UnlockBits(pictureBits);

            mask = new Bitmap(w, h, PixelFormat.Format24bppRgb);
            var maskBits = mask.LockBits(rect, ImageLockMode.WriteOnly, picture.PixelFormat);
            Marshal.Copy(maskData, 0, maskBits.Scan0, w * h * 3);
            mask.UnlockBits(maskBits);
        }

        [System.ComponentModel.DesignerCategory("")]
        class PictureConvert : AxHost
        {
            private PictureConvert() : base("") { }

            static public IPictureDisp ImageToPictureDisp(Bitmap image)
            {
                return (IPictureDisp)GetIPictureDispFromPicture(image);
            }
        }
    }
}

File: TemplateVB/Addin/AddinCommandBars.vb

Imports System.Collections.Generic
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.Linq
Imports System.Runtime.InteropServices
Imports System.Windows.Forms
Imports Microsoft.Office.Core
Imports stdole

Partial Public Class AddinUI
    Private _toolbarName As String

    Private ReadOnly _commands As New List(Of String)()

    ''' 
    ''' Constructs the UI manager
    ''' 
    Public Sub StartupCommandBars(toolbarName As String, commands As IEnumerable(Of String))
        _toolbarName = toolbarName
        _commands.AddRange(commands)

        AddHandler ThisAddIn.Application.VisioIsIdle, AddressOf ApplicationIdle
        UpdateCommandBars()
    End Sub

    Public Sub ShutdownCommandBars()
        RemoveHandler ThisAddIn.Application.VisioIsIdle, AddressOf ApplicationIdle
    End Sub

    Private _updateRequest As Boolean

    Sub UpdateCommandBars()
        _updateRequest = True
    End Sub

    Private Sub ApplicationIdle(app As Object)
        If Not _updateRequest Then
            Return
        End If

        _updateRequest = False

        UpdateToolbar()
    End Sub

    Private ReadOnly _buttons As New Dictionary(Of String, CommandBarButton)()

    Private Shared Function FindCommandBar(cbs As CommandBars, name As String) As CommandBar
        Return cbs.Cast(Of CommandBar)().FirstOrDefault(Function(cb) cb.Name = name)
    End Function

    Private Sub UpdateToolbar()
        Dim cbs = DirectCast(ThisAddIn.Application.CommandBars, CommandBars)

        Dim cb = If(FindCommandBar(cbs, _toolbarName), cbs.Add(_toolbarName))
        cb.Visible = True

        For Each id In _commands
            InstallButton(cb, id)
        Next
    End Sub

    Private Sub InstallButton(cb As CommandBar, id As String)
        Dim thisButton As CommandBarButton = Nothing
        _buttons.TryGetValue(id, thisButton)

        ' Recreate the button, otherwise it's state may be broken in Visio in some cases
        If thisButton IsNot Nothing Then
            RemoveHandler thisButton.Click, AddressOf CommandBarButtonClicked
            _buttons.Remove(id)
            Marshal.ReleaseComObject(thisButton)
        End If

        Dim button = If(DirectCast(cb.FindControl(Tag:=id), CommandBarButton), DirectCast(cb.Controls.Add(MsoControlType.msoControlButton), CommandBarButton))

        button.Enabled = ThisAddIn.IsCommandEnabled(id)

        Dim checkState = ThisAddIn.IsCommandChecked(id)
        button.State = If(checkState, MsoButtonState.msoButtonDown, MsoButtonState.msoButtonUp)

        button.Tag = id
        button.Caption = ThisAddIn.GetCommandLabel(id)
        SetCommandBarButtonImage(button, id)

        AddHandler button.Click, AddressOf CommandBarButtonClicked

        _buttons.Add(id, button)
    End Sub

    Private Sub CommandBarButtonClicked(ctrl As CommandBarButton, ByRef cancelDefault As Boolean)
        ThisAddIn.OnCommand(ctrl.Tag)
    End Sub

    Private Sub SetCommandBarButtonImage(button As CommandBarButton, id As String)
        Dim image = ThisAddIn.GetCommandBitmap(id & "_sm")
        If image Is Nothing Then
            Return
        End If

        Dim picture As Bitmap = Nothing
        Dim mask As Bitmap = Nothing
        BitmapToPictureAndMask(image, picture, mask)

        button.Picture = PictureConvert.ImageToPictureDisp(picture)
        button.Mask = PictureConvert.ImageToPictureDisp(mask)
    End Sub

    Public Shared Sub BitmapToPictureAndMask(bm As Bitmap, ByRef picture As Bitmap, ByRef mask As Bitmap)
        Dim w = bm.Width
        Dim h = bm.Height

        Dim pictureData = New Byte(3 * w * h - 1) {}
        Dim maskData = New Byte(3 * w * h - 1) {}

        Dim bmBits = bm.LockBits(New Rectangle(0, 0, w, h), ImageLockMode.[ReadOnly], PixelFormat.Format32bppArgb)
        Dim bits = New Byte(4 * w * h - 1) {}
        Marshal.Copy(bmBits.Scan0, bits, 0, 4 * w * h)
        bm.UnlockBits(bmBits)

        For y = 0 To h - 1
            For x = 0 To w - 1
                Dim srcIdx = (x + y * w) * 4
                Dim dstIdx = (x + y * w) * 3

                pictureData(dstIdx + 0) = bits(srcIdx + 0)
                pictureData(dstIdx + 1) = bits(srcIdx + 1)
                pictureData(dstIdx + 2) = bits(srcIdx + 2)

                Dim t = If((bits(srcIdx + 3) < 128), CByte(255), CByte(0))

                maskData(dstIdx + 0) = t
                maskData(dstIdx + 1) = t
                maskData(dstIdx + 2) = t
            Next
        Next

        Dim rect = New Rectangle(0, 0, w, h)

        picture = New Bitmap(w, h, PixelFormat.Format24bppRgb)
        Dim pictureBits = picture.LockBits(rect, ImageLockMode.[WriteOnly], picture.PixelFormat)
        Marshal.Copy(pictureData, 0, pictureBits.Scan0, w * h * 3)
        picture.UnlockBits(pictureBits)

        mask = New Bitmap(w, h, PixelFormat.Format24bppRgb)
        Dim maskBits = mask.LockBits(rect, ImageLockMode.[WriteOnly], picture.PixelFormat)
        Marshal.Copy(maskData, 0, maskBits.Scan0, w * h * 3)
        mask.UnlockBits(maskBits)
    End Sub

    Private Class PictureConvert
        Inherits AxHost
        Private Sub New()
            MyBase.New("")
        End Sub

        Public Shared Function ImageToPictureDisp(image As Bitmap) As IPictureDisp
            Return DirectCast(GetIPictureDispFromPicture(image), IPictureDisp)
        End Function
    End Class
End Class